File: cli/internal/excalidraw/link_test.go
package excalidraw_test

import (
	"testing"

	"obsidian-excalidraw-downloader/internal/excalidraw"
)

func TestParseAll_JsonLink(t *testing.T) {
	links := excalidraw.ParseAll("https://excalidraw.com/#json=abc123,key_abc-123", false)
	if len(links) != 1 {
		t.Fatalf("expected 1 link, got %d", len(links))
	}
	if links[0].Kind != excalidraw.JSON {
		t.Errorf("expected json, got %s", links[0].Kind)
	}
	if links[0].ID != "abc123" {
		t.Errorf("expected abc123, got %s", links[0].ID)
	}
	if links[0].Key != "key_abc-123" {
		t.Errorf("expected key_abc-123, got %s", links[0].Key)
	}
}

func TestParseAll_RoomLink(t *testing.T) {
	links := excalidraw.ParseAll("https://excalidraw.com/#room=xyz789,key_xyz-789", false)
	if len(links) != 1 {
		t.Fatalf("expected 1 link, got %d", len(links))
	}
	if links[0].Kind != excalidraw.Room {
		t.Errorf("expected room, got %s", links[0].Kind)
	}
	if links[0].ID != "xyz789" {
		t.Errorf("expected xyz789, got %s", links[0].ID)
	}
	if links[0].Key != "key_xyz-789" {
		t.Errorf("expected key_xyz-789, got %s", links[0].Key)
	}
}

func TestParseAll_Empty(t *testing.T) {
	links := excalidraw.ParseAll("", false)
	if len(links) != 0 {
		t.Fatalf("expected 0 links, got %d", len(links))
	}
}

func TestParseAll_Deduplicate(t *testing.T) {
	content := "https://excalidraw.com/#json=abc123,key_abc-123 and again https://excalidraw.com/#json=abc123,key_abc-123"
	links := excalidraw.ParseAll(content, false)
	if len(links) != 1 {
		t.Fatalf("expected 1 unique link, got %d", len(links))
	}
}

func TestParseAll_AnnotatedLink(t *testing.T) {
	content := "https://excalidraw.com/#json=abc123,key_abc-123 ([[excalidraw-abc123.excalidraw|local copy 01/01/2026]])"
	links := excalidraw.ParseAll(content, false)
	if len(links) != 0 {
		t.Fatalf("expected 0 links (already annotated), got %d", len(links))
	}
	links = excalidraw.ParseAll(content, true)
	if len(links) != 1 {
		t.Fatalf("expected 1 link in refresh mode, got %d", len(links))
	}
}

func TestParseAll_EmptyContent(t *testing.T) {
	links := excalidraw.ParseAll("just some text without any excalidraw links", false)
	if len(links) != 0 {
		t.Fatalf("expected 0 links, got %d", len(links))
	}
}

File: src/__tests__/link.test.ts
import { describe, it, expect } from 'vitest';
import { parseAll } from '../link';

describe('parseAll', () => {
  it('extracts json link kind, id, key', () => {
    const links = parseAll('https://excalidraw.com/#json=abc123,key_abc-123', false);
    expect(links).toHaveLength(1);
    expect(links[0].kind).toBe('json');
    expect(links[0].id).toBe('abc123');
    expect(links[0].key).toBe('key_abc-123');
  });

  it('extracts room link kind, id, key', () => {
    const links = parseAll('https://excalidraw.com/#room=xyz789,key-xyz', false);
    expect(links).toHaveLength(1);
    expect(links[0].kind).toBe('room');
    expect(links[0].id).toBe('xyz789');
    expect(links[0].key).toBe('key-xyz');
  });

  it('returns empty array for content without links', () => {
    expect(parseAll('just some text', false)).toHaveLength(0);
  });

  it('deduplicates multiple same URLs', () => {
    const url = 'https://excalidraw.com/#json=abc,key';
    const links = parseAll(`${url}\n${url}`, false);
    expect(links).toHaveLength(1);
  });

  it('marks already-annotated links as refresh', () => {
    const links = parseAll('([[excalidraw-abc.excalidraw|local copy 01/01/2026]]) https://excalidraw.com/#json=abc,key', false);
    expect(links).toHaveLength(1);
    expect(links[0].refresh).toBe(true);
  });

  it('handles key with underscores and hyphens', () => {
    const links = parseAll('https://excalidraw.com/#json=abc,key_with-123', false);
    expect(links).toHaveLength(1);
    expect(links[0].key).toBe('key_with-123');
  });

  it('returns empty for empty content', () => {
    expect(parseAll('', false)).toHaveLength(0);
  });
});
